### ISSUE TITLE
Bernoulli requires the response to be in {0,1}...

### ISSUE THREAD

**Issue by harrysouthworth** [gbm-developers/gbm3#6]
_From @Mullefa on November 23, 2014 17:12_

... and similarly for other distributions. Whilst this doesn't affect the functionality of the package, it is a bit inconvenient. 

Could the requirement be lowered to factors with two levels, and implicit conversion done within `gbm.fit()` (or whatever function from within the package is being used)? 

Likewise, in the corresponding`predict()` method, could a factor be returned instead of `0`'s and `1`'s?

_Copied from original issue: harrysouthworth/gbm#29_

**Comment by harrysouthworth** [comment]
_From @baek86 on February 8, 2015 16:15_

+++

**Comment by JackStat** [comment]
hmm.  I tried to just convert a y with 2 levels to a numeric 0,1 in gbm.fit but then this warning started popping up

```
Some terminal node predictions were excessively large for Bernoulli and have been capped at 1.0. Likely due to a feature that separates the 0/1 outcomes. Consider reducing shrinkage parameter.
```

Any ideas on why that could be happening?

**Comment by gregridgeway** [comment]
It means exactly what it says. In some datasets after some iterations it's possible to end up with a really large terminal node prediction that causes numerical instability. In most cases I've seen this happen when there is a feature that nearly separate the 0s and the 1s. To reduce the effects of this the terminal node will be capped at 1 and then carry on to the next iteration. Reducing the shrinkage parameter can reduce the risk of this happening.

Greg

**Comment by JackStat** [comment]
@gregridgeway I should have been more descriptive.  I created this function.

```
checkY <- function(y){

    FactorsY <- is.factor(y)
    nLevelsY <- nlevels(y)

    if(FactorsY & nLevelsY == 2){
      Y = as.numeric(y == levels(y)[2])
    }

    return(Y)

  }
```

Then towards the top of gbm.fit I added this bit

```
oldy <- y
y <- checkY(oldy)
```

But when I ran all of the tests I ended up with the error described above.  It seemed a little odd and I am not sure how it is happening.  If run through the code inside the function interactively everything seems to work fine.

**Comment by harrysouthworth** [comment]
Like Greg said, it's not a problem, it's a feature of your data. In the literature, it's often called "separation". Some predicted probabilities are 0 or 1, either because of a quirk in the training data, or because they really are (like sex as a predictor would separate cases of prostate cancer as a response).

glm with family="binomial" issues similar warnings in such settings.

**Comment by Neil-Schneider** [comment]
Bernoulli uses logit as the link function. As such, predictions in nodes are on a range between negative inifinity and inifinity. Negative inifinity translates into a prediction of zero probabablitity and inifinity translate to 100% probabablitity. If you have a feature which gives complete or near complete separation, the logit values are excessively large.  These large values, even when the predictions are shrunk by GBM, will still overwhelm the entire model. To prevent this, the predicted values on the logit scale are capped at -1.0 and 1.0.

**Comment by JackStat** [comment]
I completely understand what trips the error message.  I am saying that it is getting tripped for seemingly no reason after I try to add some code to close this issue. I will do a pull request so you guys can see the issue.

**Comment by pdmetcalfe** [comment]
Can we get this down to a minimal test case where one does the one thing and the other does the other?

``` r
fred <- sample(c(0, 1), size=1234, replace=TRUE)
bob <- factor(ifelse(fred==1, "Yes", "No"), levels=c("No", "Yes"))

set.seed(5678)
gbm(fred ~ something, dist="bernoulli")
set.seed(5678)
gbm(bob ~ something, dist="bernoulli")
```

**Comment by JackStat** [comment]
Great point!  Ok so both the numeric and factor versions fail so maybe this is an entirely different issue.  2 random uniform variables that are independent of y should not trip that error.

Update:
I just confirmed that there warning is a separate issue. On the cran version the error does not occur but it does occur with the current master branch and with the pull request that I am working on.

**Comment by pdmetcalfe** [comment]
Not surprised: cran GBM doesn't check this. It went in at 8f9480373fa62272de105a4b2d1dd12661284a16.

**Comment by Neil-Schneider** [comment]
Is -1.0 and 1.0 too restrictive for capping?  Should it be capped more at the edges of the logit distribution (-7, 7)?

**Comment by gregridgeway** [comment]
I'm open to experimentation, but I wouldn't go larger than 1. This doesn't cap the gbm prediction at 1. It caps any individual terminal node from adding more than 1 to the prediction. Subsequent iterations can add more to it. Once a single node in a single iteration hits 7 then subsequent iterations are likely to blow up. Capping at 1 for a node mitigates that risk. Again, open to other strategies and exploration of this.

Greg

Greg

**Comment by JackStat** [comment]
This issue is closed with commit 983b914
